lib/LanguageServerRename/Model/Renamer/MemberRenamer.php

```diff
@@ -0,0 +1,182 @@
+use Microsoft\PhpParser\Token;
+use Phpactor\Extension\LanguageServerRename\Model\NodeUtils;
+use Phpactor\Extension\LanguageServerRename\Model\RenameLocationGroup;
+use Phpactor\Extension\LanguageServerRename\Model\RenameLocationsProvider;
+use Phpactor\Extension\LanguageServerRename\Model\RenameResult;
+use Phpactor\Extension\LanguageServerRename\Model\Renamer;
+use Phpactor\TextDocument\ByteOffset;
+use Phpactor\TextDocument\ByteOffsetRange;
+use Phpactor\TextDocument\Location;
+use Phpactor\TextDocument\TextDocument;
+use Phpactor\TextDocument\TextDocumentLocator;
+use Phpactor\TextDocument\TextDocumentUri;
+use Phpactor\TextDocument\TextEdit;
+use Phpactor\TextDocument\TextEdits;
+
+class MemberRenamer implements Renamer
+{
+    /** @var Parser */
+    private $parser;
+    /** @var NodeUtils */
+    private $nodeUtils;
+    /** @var TextDocumentLocator */
+    private $locator;
+    /** @var RenameLocationsProvider */
+    private $locationsProvider;
+
+    public function __construct(
+        Parser $parser,
+        NodeUtils $nodeUtils,
+        TextDocumentLocator $locator,
+        RenameLocationsProvider $locationsProvider
+    ) {
+        $this->parser = $parser;
+        $this->nodeUtils = $nodeUtils;
+        $this->locator = $locator;
+        $this->locationsProvider = $locationsProvider;
+    }
+
+    private function getValidNode(TextDocument $textDocument, ByteOffset $offset): ?Node
+    {
+        $rootNode = $this->parser->parseSourceFile((string)$textDocument);
+        $node = $rootNode->getDescendantNodeAtPosition($offset->toInt());
+        
+        if (
+            $node instanceof MethodDeclaration
+            || ($node instanceof ScopedPropertyAccessExpression && $node->memberName instanceof Token)
+            || ($node instanceof Variable && $node->getFirstAncestor(PropertyDeclaration::class))
+            || ($node instanceof Variable && $node->getFirstAncestor(ScopedPropertyAccessExpression::class))
+            || $node instanceof ConstElement
+            || $node instanceof ClassConstDeclaration
+            || ($node instanceof MemberAccessExpression && $node->memberName instanceof Token)
+        ) {
+            return $node;
+        }
+        
+        return null;
+    }
+
+    public function prepareRename(TextDocument $textDocument, ByteOffset $offset): ?ByteOffsetRange
+    {
+        if (($node = $this->getValidNode($textDocument, $offset)) !== null) {
+            [ $token ] = $this->getNodeNameTokens($node);
+            return $this->nodeUtils->getTokenNameRange($token, (string)$textDocument);
+        }
+        
+        return null;
+    }
+    /**
+     * {@inheritDoc}
+     */
+    public function rename(TextDocument $textDocument, ByteOffset $offset, string $newName): Generator
+    {
+        if (($node = $this->getValidNode($textDocument, $offset)) === null) {
+            return null;
+        }
+
+        [ $token ] = $this->getNodeNameTokens($node);
```

Comment:
this should return a single Token probably.

Reply:
We need the array. Property declarations come like this:
```php
<>property string $prop1, $prop2;
```
From this point there are `$prop1` and `$prop2`.

Reply:
right, but in this class we only use a single token, the first one?

Reply:
Yes? Wait, which method are you referring to that should return a single token?

Reply:
`getNodeNameTokens` => `getNodeNameToken`

Reply:
But it can't, as I have demonstrated above :-)

Reply:
Why would we want multiple tokens though? we can only rename one property at a time?

Reply:
and here you only use one....

Reply:
if you are on the point of a property, then what are you renaming? you need to be on the point of a specific property name or?

Reply:
This is where the reference location is. So yes, I am always using just one, but I don't know which one until I examine all the names at that location:
```php
<>property string $prop1, $prop2;
```
Nearly all of the ambiguity is caused by reference locations at the start of the line/expression:
* `for` statements
* property declarations
* constant declarations

Reply:
right i see., sounds like that's what we should do then finally - I could have a look into that.

Reply:
### Property and constant declarations
The only thing I am not sure about is property and constant declarations. In the case below, it is obvious where we want the location:
```php
<>private string $prop1;
```
But in the case of `$prop2` below, I am not sure we don't want it in the same location. 
```php
private string $prop1, $prop2;
```
We would, for example, add an optional `index` to the location. In the case above, it could be `$index=0;` for `$prop1` and `$index=2` for `$prop2`.

### Foreach
Here things are clear, just in front of the variable would be nice.

### List deconstruction
We've missed the case with multi-level arrays (the definition does not work there):
```php
[ [ $var 2 ] ] = function();
```

Reply:
i'm confused -- for `private string $prop1, $prop2` we would have the location as `private string $prop1, <>$prop2)`. Ideally also with a range: `<>$prop2</>`

Reply:
Possibly could add some options to the API determining how this works.

Reply:
Anyway, can see the issue now :+1: maybe your option is the one.

Reply:
PHPStorm will do it on the var `<>$prop1` so I think we should update the ref finder to do that at least (if not to include the entire range)

Reply:
> For private string $prop1, $prop2 we would have the location as private string $prop1, <>$prop2). Ideally also with a range: <>$prop2</>

Yea, that probably is the simplest solution. I am inclined to agree. I wonder if we can check C#/Java behaviour... I'll probably do that and write back.

Reply:
In C# it selects the name and the cursor is at the end:
```c#
static string _sourcePath<> = @"E:\ERA\Source\";
```

Reply:
yep, it returns the range - i'll see if we can do the same.